### ISSUE TITLE
Invalid detaching of treesitter plugin on buffer reload

### ISSUE THREAD

**Issue by LiadOz** [nvim-treesitter/nvim-treesitter#4187]
### Describe the bug

Treesitter plugins attach detach behavior does not work correctly when a buffer gets reloaded multiple times. 

### To Reproduce

1. Create a treesitter plugin (or use any other plugin)
2. In the internal.lua file, add the following prints to M.attach and M.detach:
```lua
-- in M.attach()
print('------------------- running attach')
-- in M.detach()
print('------------------- running detach')
```
3. Open a buffer and reload it twice (one way to reload a buffer is to change the content of an open file from outside neovim)

### Expected behavior

By checking messages I get:
------------------- running attach
------------------- running detach
------------------- running attach
------------------- running attach

Expected logs:
------------------- running attach
------------------- running detach
------------------- running attach
------------------- running detach
------------------- running attach

### Output of `:checkhealth nvim-treesitter`

```markdown
==============================================================================
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- WARNING `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
- WARNING `node` executable not found (only needed for :TSInstallFromGrammar, not required for :TSInstall)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "5.15.79.1-microsoft-standard-WSL2",
  sysname = "Linux",
  version = "#1 SMP Wed Nov 23 01:01:46 UTC 2022"
} ~

Parser/Features         H L F I J
  - python              ✓ ✓ ✓ ✓ ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - tsx                 ✓ ✓ ✓ ✓ ✓
  - html                ✓ ✓ ✓ ✓ ✓
  - nvimdap             ✓ . . . ✓
  - cpp                 ✓ ✓ ✓ ✓ ✓
  - vim                 ✓ ✓ ✓ . ✓
  - help                ✓ . . . ✓
  - typescript          ✓ ✓ ✓ ✓ ✓
  - query               ✓ ✓ ✓ ✓ ✓
  - rust                ✓ ✓ ✓ ✓ ✓
  - javascript          ✓ ✓ ✓ ✓ ✓
  - lua                 ✓ ✓ ✓ ✓ ✓
  - scss                ✓ . ✓ ✓ .

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~
```

### Output of `nvim --version`

```text
NVIM v0.9.0-dev-da70c3940
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g  -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fno-common -fdiagnostics-color=always -fstack-protector-strong -DNVIM_UNIBI_HAS_VAR_FROM -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DMIN_LOG_LEVEL=3 -I/home/ozonzono/projects/nvim_plugins/neovim/.deps/usr/include/luajit-2.1 -I/usr/include -I/home/ozonzono/projects/nvim_plugins/neovim/.deps/usr/include -I/home/ozonzono/projects/nvim_plugins/neovim/build/src/nvim/auto -I/home/ozonzono/projects/nvim_plugins/neovim/build/include -I/home/ozonzono/projects/nvim_plugins/neovim/build/cmake.config -I/home/ozonzono/projects/nvim_plugins/neovim/src
Compiled by root@LIAD-TRIDENT

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"
```

### Additional context

The bug seems to come from the caching module, in there the on_detach is called on reload after attach is called and removes the keys in the cache relating to that buffer. Then when treesitter tries to determine whether a plugin should be detached, it doesn't see the keys in the cache and so doesn't detach the plugin.
I have submitted a pull request for the issue https://github.com/nvim-treesitter/nvim-treesitter/pull/4186
